lightning/common/pause.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+)
+
+// The implementation is based on https://github.com/golang/sync/blob/master/semaphore/semaphore.go
+
+// Pauser is a type which could allow multiple goroutines to wait on demand,
+// similar to a gate or traffic light.
+type Pauser struct {
+	isPaused uint32
+	mu       sync.Mutex
+	waiters  map[chan<- struct{}]struct{}
+}
+
+// NewPauser returns an initialized pauser.
+func NewPauser() *Pauser {
+	return &Pauser{
+		waiters: make(map[chan<- struct{}]struct{}, 32),
+	}
+}
+
+func (p *Pauser) set(isPaused bool) {
+	var newPaused uint32
+	if isPaused {
+		newPaused = 1
+	}
+
+	oldPaused := atomic.SwapUint32(&p.isPaused, newPaused)
+	if oldPaused == 0 || isPaused {
+		return
+	}
+
+	// extract all waiters, then notify them we changed from "Paused" to "Not Paused".
+	p.mu.Lock()
+	allWaiters := p.waiters
+	p.waiters = make(map[chan<- struct{}]struct{}, len(allWaiters))
+	p.mu.Unlock()
+
+	for waiter := range allWaiters {
+		close(waiter)
+	}
+}
+
+// Pause causes all calls to Wait() to block.
+func (p *Pauser) Pause() {
+	p.set(true)
+}
+
+// Resume causes all calls to Wait() to continue.
+func (p *Pauser) Resume() {
+	p.set(false)
+}
+
+// IsPaused gets whether the current state is paused or not.
+func (p *Pauser) IsPaused() bool {
+	return atomic.LoadUint32(&p.isPaused) != 0
+}
+
+// Wait blocks the current goroutine if the current state is paused, until the
+// pauser itself is resumed at least once.
+//
+// If `ctx` is done, this method will also unblock immediately, and return the
+// context error.
+func (p *Pauser) Wait(ctx context.Context) error {
```

Comment:
🤔 may need to introduce an intermediate "processing" state. Lemme think

Reply:
So I intend to replace the entire mutex by the atomic variable alone. WDYT

```diff
diff --git a/lightning/common/pause.go b/lightning/common/pause.go
index 26b52d1..b8ce8a8 100644
--- a/lightning/common/pause.go
+++ b/lightning/common/pause.go
@@ -15,17 +15,22 @@ package common
 
 import (
        "context"
-       "sync"
+       "runtime"
        "sync/atomic"
 )
 
+const (
+       pauseStateRunning uint32 = iota
+       pauseStatePaused
+       pauseStateLocked
+)
+
 // The implementation is based on https://github.com/golang/sync/blob/master/semaph
 
 // Pauser is a type which could allow multiple goroutines to wait on demand,
 // similar to a gate or traffic light.
 type Pauser struct {
        isPaused uint32
-       mu       sync.Mutex
        waiters  map[chan<- struct{}]struct{}
 }
 
@@ -42,16 +47,25 @@ func (p *Pauser) set(isPaused bool) {
                newPaused = 1
        }
 
-       oldPaused := atomic.SwapUint32(&p.isPaused, newPaused)
-       if oldPaused == 0 || isPaused {
-               return
+       for {
+               oldPaused := atomic.LoadUint32(&p.isPaused)
+               if oldPaused == newPaused {
+                       return
+               } 
+               if isPaused && atomic.CompareAndSwapUint32(&p.isPaused, 0, 1) {
+                       return
+               }
+               if !isPaused && atomic.CompareAndSwapUint32(&p.isPaused, 1, 2) {
+                       break
+               }
+               runtime.Gosched()
        }
 
        // extract all waiters, then notify them we changed from "Paused" to "Not Pa
-       p.mu.Lock()
        allWaiters := p.waiters
        p.waiters = make(map[chan<- struct{}]struct{}, len(allWaiters))
-       p.mu.Unlock()
+
+       atomic.StoreUint32(&p.isPaused, 0)
 
        for waiter := range allWaiters {
                close(waiter)
@@ -79,23 +93,34 @@ func (p *Pauser) IsPaused() bool {
 // If `ctx` is done, this method will also unblock immediately, and return the
 // context error.
 func (p *Pauser) Wait(ctx context.Context) error {
-       if !p.IsPaused() {
-               return nil
+       for {
+               oldPaused := atomic.LoadUint32(&p.isPaused)
+               if oldPaused == 0 {
+                       return nil
+               }
+               if atomic.CompareAndSwapUint32(&p.isPaused, 1, 2) {
+                       break
+               }
+               runtime.Gosched()
        }
 
        waiter := make(chan struct{})
-
-       p.mu.Lock()
        p.waiters[waiter] = struct{}{}
-       p.mu.Unlock()
+
+       atomic.StoreUint32(&p.isPaused, 1)
 
        select {
        case <-ctx.Done():
                err := ctx.Err()
-               p.mu.Lock()
-               delete(p.waiters, waiter)
-               p.mu.Unlock()
-               return err
+               for {
+                       oldPaused := atomic.LoadUint32(&p.isPaused)
+                       if oldPaused != 2 && atomic.CompareAndSwapUint32(&p.isPaused
+                               delete(p.waiters, waiter)
+                               atomic.StoreUint32(&p.isPaused, oldPaused)
+                               return err
+                       }
+                       runtime.Gosched()
+               }
 
        case <-waiter:
                return nil
```

Reply:
look complicate, and note runtime.Gosched() yields the processor, it does not suspend the current goroutine, I think it may take near one CPU to run an empty loop.

IMHO, just lock when you check `p.isPaused` and update `p.waiters` if need, don't use atomic for isPaused. the invariant is you need to put your waiter into `p.waiters` if `p.isPaused = 1`.
same in the set just lock while checking p.isPaused and update p.waiters.

Reply:
Using lock is like this:

```diff
diff --git a/lightning/common/pause.go b/lightning/common/pause.go
index 26b52d1..ef8e352 100644
--- a/lightning/common/pause.go
+++ b/lightning/common/pause.go
@@ -16,7 +16,6 @@ package common
 import (
        "context"
        "sync"
-       "sync/atomic"
 )
 
 // The implementation is based on https://github.com/golang/sync/blob/master/semaph
@@ -24,8 +23,8 @@ import (
 // Pauser is a type which could allow multiple goroutines to wait on demand,
 // similar to a gate or traffic light.
 type Pauser struct {
-       isPaused uint32
        mu       sync.Mutex
+       isPaused bool
        waiters  map[chan<- struct{}]struct{}
 }
 
@@ -37,18 +36,14 @@ func NewPauser() *Pauser {
 }
 
 func (p *Pauser) set(isPaused bool) {
-       var newPaused uint32
-       if isPaused {
-               newPaused = 1
-       }
-
-       oldPaused := atomic.SwapUint32(&p.isPaused, newPaused)
-       if oldPaused == 0 || isPaused {
-               return
-       }
-
        // extract all waiters, then notify them we changed from "Paused" to "Not Pa
        p.mu.Lock()
+       oldPaused := p.isPaused
+       p.isPaused = isPaused
+       if !oldPaused || isPaused {
+               p.mu.Unlock()
+               return
+       }
        allWaiters := p.waiters
        p.waiters = make(map[chan<- struct{}]struct{}, len(allWaiters))
        p.mu.Unlock()
@@ -70,7 +65,10 @@ func (p *Pauser) Resume() {
 
 // IsPaused gets whether the current state is paused or not.
 func (p *Pauser) IsPaused() bool {
-       return atomic.LoadUint32(&p.isPaused) != 0
+       p.mu.Lock()
+       isPaused := p.isPaused
+       p.mu.Unlock()
+       return isPaused
 }
 
 // Wait blocks the current goroutine if the current state is paused, until the
@@ -79,13 +77,12 @@ func (p *Pauser) IsPaused() bool {
 // If `ctx` is done, this method will also unblock immediately, and return the
 // context error.
 func (p *Pauser) Wait(ctx context.Context) error {
-       if !p.IsPaused() {
+       p.mu.Lock()
+       if !p.isPaused {
+               p.mu.Unlock()
                return nil
        }
-
        waiter := make(chan struct{})
-
-       p.mu.Lock()
        p.waiters[waiter] = struct{}{}
        p.mu.Unlock()
```

This is 5x slower in the common case: calling Wait() while not Paused:

```
// the mutex solution
PASS: pause_test.go:144: pauseSuite.BenchmarkWaitCtxCanceled     5000000               316 ns/op
PASS: pause_test.go:136: pauseSuite.BenchmarkWaitNoOp   100000000               26.1 ns/op
// the atomic solution
PASS: pause_test.go:144: pauseSuite.BenchmarkWaitCtxCanceled     5000000               311 ns/op
PASS: pause_test.go:136: pauseSuite.BenchmarkWaitNoOp   500000000                4.31 ns/op
```

As Wait() is called quite frequently so I'd like to ensure is it as fast as possible.